pkg/controller/control_test.go

```diff
@@ -86,3 +91,208 @@ func (s *TestSuite) TestDetermineCorrectVolumeSize(c *C) {
+		}
+	}
+	testsets = append(testsets, testCase{buf: buf, off: off, expectedData: expectedData})
+
+	// Test case #2
+	buf = makeByteSliceWithInitialData(512, newVal)
+	off = 4096 - 512
+	expectedData = makeByteSliceWithInitialData(dataLength, writeSourceInitVal)
+	for i := 0; i < len(expectedData); i++ {
+		switch {
+		case i < 4096-512:
+			expectedData[i] = readSourceInitVal
+		case i < 4096:
+			expectedData[i] = newVal
+		}
+	}
+	testsets = append(testsets, testCase{buf: buf, off: off, expectedData: expectedData})
+
+	// Test case #3
+	buf = makeByteSliceWithInitialData(4096+1024, newVal)
+	off = 4096 - 512
+	expectedData = makeByteSliceWithInitialData(dataLength, writeSourceInitVal)
+	for i := 0; i < len(expectedData); i++ {
+		switch {
+		case i < 4096-512:
+			expectedData[i] = readSourceInitVal
+		case i < 4096-512+4096+1024:
+			expectedData[i] = newVal
+		case i < 4096*3:
+			expectedData[i] = readSourceInitVal
+		}
+	}
+	testsets = append(testsets, testCase{buf: buf, off: off, expectedData: expectedData})
+
+	// Test case #4
+	buf = makeByteSliceWithInitialData(4096, newVal)
+	off = 4096
+	expectedData = makeByteSliceWithInitialData(dataLength, writeSourceInitVal)
+	for i := 0; i < len(expectedData); i++ {
+		switch {
+		case i < 4096:
+			continue
+		case i < 4096+4096:
+			expectedData[i] = newVal
+		}
+	}
+	testsets = append(testsets, testCase{buf: buf, off: off, expectedData: expectedData})
+
+	// Test case #5
+	buf = makeByteSliceWithInitialData(4096*2, newVal)
+	off = 4096
+	expectedData = makeByteSliceWithInitialData(dataLength, writeSourceInitVal)
+	for i := 0; i < len(expectedData); i++ {
+		switch {
+		case i < 4096:
+			continue
+		case i < 4096*3:
+			expectedData[i] = newVal
+		}
+	}
+	testsets = append(testsets, testCase{buf: buf, off: off, expectedData: expectedData})
+
+	// Test case #6
+	buf = makeByteSliceWithInitialData(4096+512, newVal)
+	off = int64(dataLength - 4096 - 512)
+	expectedData = makeByteSliceWithInitialData(dataLength, writeSourceInitVal)
+	for i := 0; i < len(expectedData); i++ {
+		switch {
+		case i < 4096*2:
+			continue
+		case i < 4096*4-4096-512:
+			expectedData[i] = readSourceInitVal
+		case i < 4096*4:
+			expectedData[i] = newVal
+		}
+	}
+	testsets = append(testsets, testCase{buf: buf, off: off, expectedData: expectedData})
+
+	// Test case #7
+	buf = makeByteSliceWithInitialData(512, newVal)
+	off = int64(dataLength - 512)
+	expectedData = makeByteSliceWithInitialData(dataLength, writeSourceInitVal)
+	for i := 0; i < len(expectedData); i++ {
+		switch {
+		case i < dataLength-4096:
+			continue
+		case i < dataLength-512:
+			expectedData[i] = readSourceInitVal
+		case i < dataLength:
+			expectedData[i] = newVal
+		}
+	}
+	testsets = append(testsets, testCase{buf: buf, off: off, expectedData: expectedData})
+
+	// Test case #8
+	buf = makeByteSliceWithInitialData(0, newVal)
+	off = 4096 * 3
+	expectedData = makeByteSliceWithInitialData(dataLength, writeSourceInitVal)
```

Comment:
Isn't this `readSourceInitVal`?

Reply:
I think no. the buff has 0 length. the fakeWriter.source shouldn't be touched so expected data remain as writeSourceInitVal

Reply:
Oh Yes. I misunderstood `writeSource` a little bit

Reply:
But you are right that we would have problem in this new test case:
```
	// Test case #9
	buf = makeByteSliceWithInitialData(0, newVal)
	off = 4096 + 512
```

the fakeWriter.source shouldn't be touched therefore, I added the condition:

```
	if (bufLen == 0) || ((off%util.VolumeSectorSize == 0) && (bufLen%util.VolumeSectorSize == 0)) {
		return c.backend.WriteAt(b, off)
	}
```

LOL. unit test is very helpful